lib/sql.ml

```diff
@@ -110,13 +110,14 @@ struct
   type tyvar = Typ of t | Var of int
   let string_of_tyvar = function Typ t -> show t | Var i -> sprintf "'%c" (Char.chr @@ Char.code 'a' + i)
 
+  type agg_fun = Self | Count | Avg (* max, min | count | avg *)
```

Comment:
i was thinking i would prefer to have return type here (`Self | Return of strict int | Return of nullable float), but then realize that indeed it is count and avg are so special and if there are no other functions that behave in same way then this also makes sense. But please put comment that describes why these functions are special

Reply:
Hmm... hm hm hm, I am thinking now, and I realize that you are right about `avg`.
`Count` - should be kept, no one aggregation in the MySQL world can't be non-nullable and every-time-row-returnable, unlike count.
When I was adding `avg,` my thoughts were that we split functions for the following kinds: self, which means it returns the same type as rows have, and count (already described), and avg, which returns not self type but produces calculation and return a float.
But now I chatGPTed about which functions do calculations besides avg, and it returned to me: `STD` (I checked it and , yes, it has the same type that avg),  `VAR_POP` (same),  `GROUP_CONCAT` - okay it deserves a new constructor, `BIT_AND` (it accepts   `TINYINT`, `SMALLINT`, `MEDIUMINT`, `INT`, `BIGINT` and returns `BIGINT`, and how it should be described. Roughly: Int -> Int Nullable.

Reply:
`type agg_fun = Self | Count | Return of int | Return of floar`

Reply:
Another problem, is that it's possible to write:

```sql
SELECT AVG('string');
```

It won't raise an error in runtime. it will try to cast a string to int, but I guess it is better to restrict it.
And the second question is where to restrict?
To return something like during inferring: 🤔

```ocaml
 | Agg Avg, [_] -> 
          let (_, typ) = typeof (ResFun ((F ((Typ (nullable Float)), [(Typ (depends Int))])), params)) in 
          let typ = get_or_failwith typ in
          consider_groupping @@ typ, types
```
But this way we miss Float

Reply:
> But this way we miss Float

But we can match argument, okay better I guess to to "desugaring" carefully before calling infer.

Reply:
Okay, for now, I will check it in pattern matching during infer, but what if we will do something like typeclasses
https://github.com/andreypopp/type-systems/blob/main/algo_w/infer.ml#L8 to pick Number class, but yes not for this PR

Reply:
> It won't raise an error in runtime. it will try to cast a string to int, but I guess it is better to restrict it.

currently sqlgg doesn't do any validation and it is good for adoption. Validation will require better impl for type inference etc, and anyhow must be behind the switch to not disrupt